src/main/java/billgates/view/gui/ActionPanel.java

```diff
@@ -223,4 +213,32 @@ private void deleteEntry() {
     public JButton getDeleteEntryButton() {
         return deleteEntryButton;
     }
+
+    @Override
+    public void view(UserJoinViewModel viewModel) {
+        // If the user join successfully,
+        if (viewModel.isJoined()) {
+            // disable the signInButton, and enable the signOutButton and addEntryButton
+            this.signInButton.setEnabled(false);
+            this.signOutButton.setEnabled(true);
+            this.addEntryButton.setEnabled(true);
+
+            // the usernameField and passwordField shouldn't be editable
+            this.usernameField.setEditable(false);
+            this.passwordField.setEditable(false);
+
+            // enable importMenu
+            TopMenuBar tmb = (TopMenuBar) this.getRootPane().getJMenuBar();
+            tmb.getImportMenu().setEnabled(true);
+
+            // enable billTable
+            MainFrame mf = (MainFrame) SwingUtilities.getWindowAncestor(this);
+            BillTable bt = (BillTable) mf.getBillPanel().getBillTable();
+            bt.setVisible(true);
+            bt.setEnabled(true);
+        }
+
+        // Show a message dialog with whatever the text from the viewModel
+        JOptionPane.showMessageDialog(this.getParent(), viewModel.getReasonRejected());
```

Comment:
`this.getParent()` could be replaced by the `mainFrame` instance variable.

Also, it should only show the dialog if the user failed to join. In this case, it seems like it will pop a dialog regardless of whether the user is joined.

Reply:
Our design is to always show a dialog no matter if the user has successfully joined or failed to join.
But the dialog's messages are different:
1. If the user has successfully joined, it will notify the user that he/she has successfully joined OR has successfully registered.
2. If the user failed to join, it will also notify the user of the reason.

Reply:
Great!